### PR TITLE
Move *io* as first argument of method overloads

### LINKED ISSUE #5916: Standardize overloads for methods with IO/returning String
The idiomatic Crystal way of converting a value to a `String` is to append it to an `IO`. The most notable it `Object#to_s(io : IO)` which writes a string representation to the `IO`. When an actual `String` is required, there is an overload `Object#to_s()` which calls the IO variant wrapped in `String.build`.

Other examples are `Object#inspect`, `Base64#encode`, `Base64#decode` or `Enumerable#join` which all adapt the same technique.

Since they provide the same functionality, both methods usually have the same signature except that one returns `String` and the other accepts an additional argument `io : IO`. That's the distinctive difference between them. It is obvious that the positioning of the other arguments should be the same to avoid any friction.

If the methods also accept other arguments, there is a question as to where the `io` argument should be placed. There are different styles throughout the standard lib:

* `Time#to_s(format : String, io : IO)`
* `Int#to_s(base : Int, io : IO, upcase : Bool = false)`
* `BigRational#to_s(io : IO, base = 10)`

It doesn't make any sense to add `io` in between as it screws up argument order. At the end is better but it will cause issues when optional arguments are added to the signature (that's probably what happened with `Int#to_s`). I think it should be preferred to put `io` at the front of the other arguments. This clearly signals that this is the IO variant of the method. And additional arguments can simply be appended at the end in both variants. IMHO this is the most logical solution.

I propose to adapt "`io` in front" style throughout the standard library and encourage others to do the same. This way it would be simple to switch from a String method to the IO method by inserting `io` as first argument and you don't need to think about where to put it.

### PR DESCRIPTION
**Pull request by straight-shoota**
Fixes #5916

### CODE REVIEW

File: spec/std/int_spec.cr [hunk 1]
```diff
@@ -2,11 +2,11 @@ require "spec"
 require "big"
 
 private def to_s_with_io(num)
-  String.build { |str| num.to_s(str) }
+  String.build { |io| num.to_s(io) }
```
Comment (ysbaddaden) [inline spec/std/int_spec.cr]: Unrelated change. I also prefer the `str` idiom for `String.build`. We're building a string, not whatever io.

**Comment by ysbaddaden** [comment]
Maybe injecting the `io` in middle of arguments is a hint that some arguments should be named arguments (thus optional)?

**Comment by straight-shoota** [comment]
@ysbaddaden Could we keep the general discussion about this in #5917, please?

File: spec/std/int_spec.cr (hunk 1 above)
Comment (straight-shoota) [reply spec/std/int_spec.cr]: I can revert that.

But I always find this weird ("why is this passing a String?"): `str` is not a string but a string builder (so it would make sense to call it `builder`). Which is an IO and `io` is shorter but equally expressive. The argument in the method definition of `to_s` is also called `io`. So it makes very much sense to call this `io`.

**Comment by Sija** [comment]
@straight-shoota You meant #5916?

**Comment by straight-shoota** [comment]
Closing. There has been no resolution on this and the PR would need a huge rebase, so it would be easier to start fresh.
